src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.cs

```diff
@@ -40,24 +43,28 @@ internal VisualStudioDiagnosticAnalyzerProvider(object extensionManager, Type ty
 
         _extensionManager = extensionManager;
         _typeIExtensionContent = typeIExtensionContent;
-        _lazyAnalyzerReferences = new Lazy<ImmutableArray<(AnalyzerFileReference, string)>>(GetAnalyzerReferencesImpl);
+        _lazyAnalyzerReferences = new Lazy<ImmutableArray<(AnalyzerFileReference, string)>>(() => GetExtensionContent(AnalyzerContentTypeName).SelectAsArray(c => (new AnalyzerFileReference(c.path, AnalyzerAssemblyLoader), c.extensionId)));
+        _lazyRazorReferences = new Lazy<ImmutableArray<(string, string)>>(() => GetExtensionContent(RazorContentTypeName));
     }
 
     public ImmutableArray<(AnalyzerFileReference reference, string extensionId)> GetAnalyzerReferencesInExtensions()
         => _lazyAnalyzerReferences.Value;
 
-    private ImmutableArray<(AnalyzerFileReference reference, string extensionId)> GetAnalyzerReferencesImpl()
+    public ImmutableArray<(string path, string extensionId)> GetRazorReferencesInExtensions()
+        => _lazyRazorReferences.Value;
+
+    private ImmutableArray<(string path, string extensionId)> GetExtensionContent(string contentTypeName)
     {
         try
         {
             // dynamic is weird. it can't see internal type with public interface even if callee is
             // implementation of the public interface in internal type. so we can't use dynamic here
-            var _ = PooledDictionary<AnalyzerFileReference, string>.GetInstance(out var analyzePaths);
+            var _ = PooledDictionary<string, string>.GetInstance(out var analyzePaths);
 
-            // var enabledExtensions = extensionManager.GetEnabledExtensions(AnalyzerContentTypeName);
+            // var enabledExtensions = extensionManager.GetEnabledExtensions(contentTypeName);
             var extensionManagerType = _extensionManager.GetType();
             var extensionManager_GetEnabledExtensionsMethod = extensionManagerType.GetRuntimeMethod("GetEnabledExtensions", [typeof(string)]);
-            var enabledExtensions = (IEnumerable<object>)extensionManager_GetEnabledExtensionsMethod.Invoke(_extensionManager, [AnalyzerContentTypeName]);
+            var enabledExtensions = (IEnumerable<object>)extensionManager_GetEnabledExtensionsMethod.Invoke(_extensionManager, [contentTypeName]);
```

Comment:
> Previously this was ok, as the assemblies only contained a source generator, not any analyzers, so listing them as VSIX analyzers was essentially a NOP.

Trying to understand this. When we had

```xml
<Asset Type="Microsoft.VisualStudio.Analyzer" Path="Microsoft.CodeAnalysis.Razor.Compiler.dll" />
```

how come it wasn't loaded until we added the suppressor? Shouldn't Roslyn need to load the DLL and look inside to determine whether there is an analyzer or not?

Reply:
It was loaded once, to see if it had any analyzers etc. in it. It's then loaded again later on if it was found to contain analyzers previously. Because it didn't find anything before, we didn't load it the second time.